tensorboard/plugins/custom_scalar/polymer3/tf_custom_scalar_dashboard/tf-custom-scalar-margin-chart-card.ts

```diff
@@ -347,12 +356,23 @@ class _TfCustomScalarMarginChartCard extends LegacyElementMixin(PolymerElement)
   _logScaleActive: boolean;
 
   @property({type: Object})
-  _dataUrl: (run: string) => string = (run) => {
-    const tag = this._tagFilter;
-    return addParams(getRouter().pluginRoute('custom_scalars', '/scalars'), {
-      tag,
-      run,
-    });
+  _requestData: RequestDataCallback<RunItem, CustomScalarsDatum> = (
+    items,
+    onLoad,
+    onFinish
+  ) => {
+    const router = getRouter();
+    const baseUrl = router.pluginRoute('custom_scalars', '/scalars');
+    Promise.all(
```

Comment:
Hmm, I wonder if it makes more sense to just return the promise in this callback.

Reply:
So, there were a couple models here that I considered. You could just
replace the `onFinish` continuation with “returns `Promise<void>`”,
which I presume is what you’re suggesting. That works, but I thought it
was a bit confusing to mix callbacks and promises. The alternative that
I’d be more interested in is requiring the function to just return an
array of `Promise<{item: Item, data: Data}>`—either that or an
early-bound array of `{item: Item, data: Promise<Data>}`. In some ways
this felt a bit more restrictive because `requestData` has to explicitly
farm out promises for each item, but of course it’s equally flexible
assuming that you don’t do unexpected things like call the `onLoad`
callback multiple times. This also has a nice way to reject individual
promises, which would be a strong argument if `DataLoaderBehavior` had a
notion of failed load states.

Do you have thoughts or preferences here?

Reply:
Ok, I buy the argument. While I think the node-like APIs are a bit dated, I agree now that maybe we may not want to mix callback + Promise.
 
I definitely think we can re-evaluate `onLoadDataCallback` and `onLoadFinish`. Not sure how that will change the API structure though. Since we have plans to redo scalars, maybe I can borrow your thoughts on the API organization then.